/**
 * This file is part of the "libfnord" project
 *   Copyright (c) 2015 Paul Asmuth
 *
 * FnordMetric is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License v3.0. You should have received a
 * copy of the GNU General Public License along with this program. If not, see
 * <http://www.gnu.org/licenses/>.
 */
package zbase;

enum TablePartitionerType {
  TBL_PARTITION_TIMEWINDOW = 0;
  TBL_PARTITION_FIXED = 1;
};

enum TableStorage {
  TBL_STORAGE_COLSM = 0;
  TBL_STORAGE_STATIC = 2;
};

message TableConfig {

  /**
   * The schema for this table
   */
  required bytes schema = 1;

  /**
   * The partitioner for this table
   */
  required TablePartitionerType partitioner = 2;

  /**
   * What kind of data is stored in this table?
   */
  required TableStorage storage = 3;

  /**
   * The maximum sstable size in bytes
   */
  optional uint64 sstable_size = 5 [default = 536870912];

  /**
   * The number of shards for this table
   */
  optional uint64 num_shards = 7;

  /**
   * How often should we commit the partition to disk in microseconds (N.B.
   * that even with a high value, we can't loose data as inserts are replicated
   * before the interval is up)
   */
  optional uint64 commit_interval = 9 [default = 60000000];

  optional TimeWindowPartitionerConfig time_window_partitioner_config = 8;

  /**
   * If true, allow unauthorized inserts into this table (e.g. for the tracking
   * pixel or for events sent from a mobile device
   */
  optional bool allow_public_insert = 10 [default = true];

};

message TimeWindowPartitionerConfig {

  /**
   * The partition size in microseconds
   */
  required uint64 partition_size = 1;

};

message TableDefinition {

  required string customer = 1;

  required string table_name = 2;

  optional uint64 version = 3;

  required TableConfig config = 4;

  optional uint64 next_field_id = 5;

  repeated string tags = 6;
};

message TableDefinitionList {

  repeated TableDefinition tables = 1;

  optional uint64 version = 2;

};
